src/WishedPrice/Infrastructure/WishedPriceFactory.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+/**
+ * Copyright © OXID eSales AG. All rights reserved.
+ * See LICENSE file for license details.
+ */
+
+declare(strict_types=1);
+
+namespace OxidEsales\GraphQL\Account\WishedPrice\Infrastructure;
+
+use OxidEsales\Eshop\Application\Model\PriceAlarm;
+use OxidEsales\GraphQL\Account\WishedPrice\DataType\WishedPrice;
+use OxidEsales\GraphQL\Account\WishedPrice\Exception\WishedPriceOutOfBounds;
+use OxidEsales\GraphQL\Base\Exception\NotFound;
+use OxidEsales\GraphQL\Base\Service\Authentication;
+use OxidEsales\GraphQL\Catalogue\Currency\Infrastructure\Repository as CurrencyRepository;
+use OxidEsales\GraphQL\Catalogue\Product\DataType\Product as ProductDataType;
+use OxidEsales\GraphQL\Catalogue\Product\Exception\ProductNotFound;
+use OxidEsales\GraphQL\Catalogue\Shared\Infrastructure\Repository;
+use TheCodingMachine\GraphQLite\Types\ID;
+
+final class WishedPriceFactory
+{
+    /** @var Authentication */
+    private $authentication;
+
+    /** @var Repository */
+    private $repository;
+
+    /** @var CurrencyRepository */
+    private $currencyRepository;
+
+    public function __construct(
+        Repository $repository,
+        CurrencyRepository $currencyRepository
+    ) {
+        $this->repository         = $repository;
+        $this->currencyRepository = $currencyRepository;
+    }
+
+    public function createValidWishedPrice(
+        string $userId,
+        string $userName,
+        ID $productId,
+        string $currencyName,
+        float $price
+    ): WishedPrice {
+        $this->assertProductWishedPriceIsPossible($productId);
+        $this->assertPriceValue($price);
+
+        $currency = $this->currencyRepository->getByName($currencyName);
+
+        /** @var PriceAlarm $model */
+        $model = oxNew(PriceAlarm::class);
+        $model->assign(
+            [
+                'OXUSERID'   => $userId,
+                'OXEMAIL'    => $userName,
+                'OXARTID'    => (string) $productId->val(),
+                'OXPRICE'    => round($price, $currency->getPrecision()),
+                'OXCURRENCY' => $currency->getName(),
+            ]
+        );
+
+        return new WishedPrice($model);
+    }
+
+    /**
+     * @throws ProductNotFound
+     *
+     * @return true
+     */
+    private function assertProductWishedPriceIsPossible(ID $productId): bool
+    {
+        $id = (string) $productId->val();
+
+        try {
+            /** @var ProductDataType $product */
+            $product = $this->repository->getById($id, ProductDataType::class);
+        } catch (NotFound $e) {
+            throw ProductNotFound::byId($id);
+        }
+
+        // Throw 404 if product has wished prices disabled
+        if (!$product->getEshopModel()->isPriceAlarm()) {
+            throw ProductNotFound::byId($id);
```

Comment:
Just an opinion: a `404` status code seems wrong here, but I don't know of a better matching status code then the overly generic `400` ..

Reply:
maybe use 418 I'm a teapot ?  or 405 Method Not Allowed ?

Reply:
I guess we can just leave the `404`

Reply:
Point is: 404 give the hint that product is not available for wished price. 400 would only be some 'bad request'. Looks like we'll leave it at 404 for now.